### PR TITLE
osutil/disks: calculate the last usable LBA instead of reading it

### PR DESCRIPTION
**Pull request by valentindavid**
Installation of Ubuntu Core is usually done by writing an image onto a
medium. The size of the medium is usually bigger than the image.  It
is expected that the partition `ubuntu-data` will be sized to fill up
that medium.

Because GPT has a backup header and table at the end of the disk, it
is not possible to create a partition until the end of this disk.
For this reason the entry "last usable LBA" from the header
tells us until where we can create partitions.

For the case of Ubuntu Core, since the size of the medium is typically
bigger than the image, the backup header and table will not be at the
end of the disk when reading it. It will be fixed later when updating
the table.

For that reason, the "last usable LBA" from the image will not
be the correct one for the disk.

`sfdisk --json` used to report a calculated "last usable LBA" instead
of the effective value stored in the header. But it has been changed
since 387ac2771f1b16af00c417b226c697d08d1301fe (2.36), and now it
reports the value from the header.

For that reason we have to calculate the expected "last usable LBA"
that will be used. This is calculated from the size of the table
and the size of the disk.

While it is possible to also to `echo write | sfdisk /dev/disk`,
to fix before `sfdisk --json`, we would require to wait for
udev to treat all events.

### CODE REVIEW

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:33]: a link somewhere definitive (even just util-linux is fine) for this format would be good

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:66]: from this point below, all the error conditions should return and empty GPTHeader{} here we may be returning an structure which is valid in terms of a Go structure and has data in the fields and such, but logically is corrupted because it doesn't have the right signature, crc doesn't match etc.

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:134]: could you add some comments explaining these operations? AFAICT you're just counting how many headers are used by the gpt headers / backup headers, and then subtracting that from the total number of sectors, but it's a bit unclear to my why you are doing `(entriessize + 511) / 512` for example

**Review by anonymouse64 — COMMENTED** [review]
> While it is possible to also to echo write | sfdisk /dev/disk,
to fix before sfdisk --json, we would require to wait for
udev to treat all events.

How much have you explored this option? How would we be able to tell when udev is done processing events? I ask because we already very rough version of such logic in the gadget/install package, so it doesn't seem super unreasonable to me to have a step in the gadget/install.Run() function where before doing anything with the disk, we do `echo write | sfdisk /dev/foo` and then wait for some events to finish and then we can continue to rely on sfdisk doing all this math for us. That would save us from having to maintain this code here.

I'm not against maintaining this code here, just that the less code we have to maintain here the better, and I trust util-linux/sfdisk maintainers more than myself for example to get all the details correct.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:51]: Done.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:34]: Done.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done.

**Comment by valentindavid** [comment]
> How much have you explored this option? How would we be able to tell when udev is done processing events? I ask because we already very rough version of such logic in the gadget/install package, so it doesn't seem super unreasonable to me to have a step in the gadget/install.Run() function where before doing anything with the disk, we do `echo write | sfdisk /dev/foo` and then wait for some events to finish and then we can continue to rely on sfdisk doing all this math for us. That would save us from having to maintain this code here.

I have implemented it and ran it with the tests `tests/main/uc20-create-partitions` and `tests/main/uc20-create-partitions-encrypt`. And I needed to call `udevTrigger` after running `echo write | sfdisk --no-reread /dev/node`. Which means we need to wait for udev. And this is every time, even if the disk was not resized. And there is no real nice way to detect that. The installation is already very slow. Adding another wait is not going to help.

Also I am not totally the concurrency safety of `udevadm trigger --settle`. Could it happen that the kernel has not updated the device uevent yet, even though `ioctl(BLKRRPART)` has been called by `sfdisk`? And then later while we are reading udev properties, udev does the updates which starts by clearing all the properties before re-adding them.

I have tried things like `--no-reread --no-tell-kernel` to not trigger at all udev. That did not work. I am not sure why, but something else could have triggered a re-read of the disk. And in theory this issue could happen.

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:59]: I think it would make more sense for the logic to assume it's always either able to read a full sector's worth of data, and if it is not, then it's most likely an error and it should not assume integrity of data is valid? So instead,
```read < len(rawHeader)```

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:66]: Maybe a bit nitpicky, but it would be nice to see the validation logic be split into a seperate function, to keep the function intentions clean.

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:67]: s/magic string/the magic string/

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:75]: Nitpick, maybe 'GPT header size is smaller than the minimum valid size'

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:79]: Nitpick, suggestion: 'GPT header size is larger than the maximum supported size'

File: tests/main/uc20-create-partitions-encrypt/task.yaml
Comment (Meulengracht) [inline tests/main/uc20-create-partitions-encrypt/task.yaml:38]: Awesome!

**Review by Meulengracht — COMMENTED** [review]
Really happy to see this getting fixed! Just some small comments and nitpicks (sorry for those, feel free to ignore them)

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:59]: Technically, for us, we do not really need the sector to be complete, even though it would be weird. I suppose that could happen with an image file from a network mount. It is just a little more robust to not check against a bigger value, but what we exactly need.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:52]: Done. Though due to coupling of `rawHeader` and `header`, and we need both to verify, it does not look great.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:72]: this can be scoped as you don't need the 'crc' variable other places than inside if check:
```
if crc := crc32.ChecksumIEEE(rawHeader[:header.HeaderSize]); crc != header.HeaderCRC {
	return fmt.Errorf("GPT header CRC32 checksum failed: %v != %v", crc, header.HeaderCRC)
}
```

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:94]: this can also be scoped if I am not wrong
```
if err := verifyHeader(rawHeader[:read], header); err != nil {
	return GPTHeader{}, err
}
```

File: osutil/disks/gpt.go
Comment (Meulengracht) [inline osutil/disks/gpt.go:110]: same for scoping

**Review by Meulengracht — APPROVED** [review]
In general, minor styling changes, otherwise looks good to me

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Putting this one in one line made it difficult to read. So instead I put the code in the block which limits the scope of `crc`.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

**Review by Meulengracht — APPROVED** [review]
Thanks for the revisions, I'm happy LGTM :)

File: osutil/disks/gpt.go
Comment (mardy) [inline osutil/disks/gpt.go:117]: You can use SEEK_SET and SEEK_END: https://pkg.go.dev/os#pkg-constants

File: osutil/disks/gpt.go
Comment (mardy) [inline osutil/disks/gpt.go:142]: I'm not sure that we can hardcode 512 as the sector size here; we do have a `blockDeviceSectorSize()` function, so it would be better to use that for consistency.

**Review by mardy — APPROVED** [review]
LGTM, with some minor remarks/nitpicks :-)

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:142]: We are talking about *logical* blocks. So it is always 512. `blockDeviceSectorSize` could give us something that is not the same as logical blocks.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:76]: @Meulengracht had concerns about the scope of `crc` during his review.

File: osutil/disks/gpt.go
Comment (mardy) [reply osutil/disks/gpt.go:142]: Ah, I see. And `blockDeviceSizeInSectors()` also reports logical sectors, so this is indeed correct.

**Review by mardy — APPROVED** [review]

File: osutil/disks/disks_linux.go
Comment (anonymouse64) [inline osutil/disks/disks_linux.go:378]: we need a unit test in disks_linux_test for the case where this is successful to demonstrate we don't try sfdisk in the case where this works, currently we are only testing the case where this fails and we fall back to using sfdisk

File: osutil/disks/gpt.go
Comment (Meulengracht) [reply osutil/disks/gpt.go:76]: Yes, as Ian said, I just wanted to change it as a stylistic thing

File: osutil/disks/disks_linux.go
Comment (valentindavid) [reply osutil/disks/disks_linux.go]: Done.

File: osutil/disks/disks_linux.go
Comment (anonymouse64) [inline osutil/disks/disks_linux.go:51]: @valentindavid the mac OS build of the `snap` command no longer builds:
```
Error: osutil/disks/mockdisk.go:412:9: undefined: abstractCalculateLastUsableLBA
```

you may need to put this definition into `disks.go` instead of here in disks_linux.go

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:68]: this is clearing the Reserved field in the raw header correct? why do we need to clear it here? As per the spec it should always be zero, no? if it's not zero then the GPT isn't valid?

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:148]:
```suggestion
	//	| Backup Header          |  sectors - 1
```

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:142]: thanks for this diagram

File: osutil/disks/gpt.go
Comment (anonymouse64) [inline osutil/disks/gpt.go:145]:
```suggestion
	//	| Backup Table           |  sectors - 1 - tableSizeInSectors
```

**Review by anonymouse64 — COMMENTED** [review]
lgtm, some small nitpicks and one question about the zeroing of the Reserved field which seems unexpected to me

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go:68]: It contains the crc. But the crc is calculated with the stored crc being 0.

File: osutil/disks/gpt.go
Comment (valentindavid) [reply osutil/disks/gpt.go]: Done

File: tests/main/uc20-create-partitions/task.yaml
Comment (valentindavid) [reply tests/main/uc20-create-partitions/task.yaml]: Done

**Review by anonymouse64 — APPROVED** [review]
lgtm though I realize now before merging we should run the nested tests here too just in case

File: osutil/disks/gpt_test.go
Comment (anonymouse64) [inline osutil/disks/gpt_test.go:49]: It seems on 16.04 and 14.04, the sfdisk command is not available:
```

----------------------------------------------------------------------
FAIL: gpt_test.go:41: gptSuite.SetUpTest

gpt_test.go:49:
    c.Assert(err, IsNil)
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc0002cc380), Stderr:[]uint8(nil)} ("exit status 1")


----------------------------------------------------------------------
PANIC: gpt_test.go:178: gptSuite.TestBadCRC

... Panic: Fixture has panicked (see related PANIC)
OOPS: 46 passed, 1 FAILED, 12 MISSED
--- FAIL: Test (0.27s)
FAIL
FAIL	github.com/snapcore/snapd/osutil/disks	0.274s
```

is there anyway we can setup a gpt partition image without using sfdisk this way for testing purposes? it seems we have the following options:

* document that sfdisk is needed to run unit tests and possibly adjust build-deps
* skip this unit test when sfdisk is not available
* embed a generated header / backup header bytes into the test as a []byte variable and write them out to a file for the test

Comment (valentindavid) [reply osutil/disks/gpt_test.go]: I have put the headers as files in testdata and rebuild the image from those. There is a script that can re-generates the files.

**Review by mvo5 — APPROVED** [review]
Thank you

**Review by anonymouse64 — COMMENTED** [review]
the last changes lgtm fwiw, I do like basing the behavior on version actually

**Pull request merged**
